sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java

```diff
@@ -113,6 +153,302 @@ public void validateWindow(SqlNode windowOrId, SqlValidatorScope scope, @Nullabl
+   * @param ingestNode    The ingest node.
+   * @param tableMetadata The table metadata as stored in the catalog, if any.
+   *
+   * @return The effective granularity
+   * @throws org.apache.druid.error.DruidException indicating invalud Sql if both the ingest node and table metadata
+   * for the respective target table have no PARTITIONED BY granularity defined.
+   */
+  private Granularity getEffectiveGranularity(
+      final String operationName,
+      final DruidSqlIngest ingestNode,
+      @Nullable final DatasourceFacade tableMetadata
+  )
+  {
+    Granularity effectiveGranularity = null;
+    final Granularity ingestionGranularity = ingestNode.getPartitionedBy() != null
+        ? ingestNode.getPartitionedBy().getGranularity()
+        : null;
+    if (ingestionGranularity != null) {
+      DruidSqlParserUtils.validateSupportedGranularityForPartitionedBy(ingestNode, ingestionGranularity);
+      effectiveGranularity = ingestionGranularity;
+    } else {
+      final Granularity definedGranularity = tableMetadata == null
+          ? null
+          : tableMetadata.segmentGranularity();
+      if (definedGranularity != null) {
+        // Should already have been checked when creating the catalog entry
+        DruidSqlParserUtils.validateSupportedGranularityForPartitionedBy(null, definedGranularity);
+        effectiveGranularity = definedGranularity;
+      }
+    }
+
+    if (effectiveGranularity == null) {
+      throw InvalidSqlInput.exception(
+          "Operation [%s] requires a PARTITIONED BY to be explicitly defined, but none was found.",
+          operationName);
+    }
+
+    return effectiveGranularity;
+  }
+
+  /**
+   * Compute and validate the target type. In normal SQL, the engine would insert
+   * a project operator after the SELECT before the write to cast columns from the
+   * input type to the (compatible) defined output type. Druid doesn't work that way.
+   * In MSQ, the output the just is the input type. If the user wants to control the
+   * output type, then the user must manually insert any required CAST: Druid is not
+   * in the business of changing the type to suit the catalog.
+   * <p>
+   * As a result, we first propagate column names and types using Druid rules: the
+   * output is exactly what SELECT says it is. We then apply restrictions from the
+   * catalog. If the table is strict, only column names from the catalog can be
+   * used.
+   */
+  private RelDataType validateTargetType(
+      SqlValidatorScope scope,
+      final IdentifierNamespace insertNs,
+      SqlInsert insert,
+      RelRecordType sourceType,
+      DatasourceFacade tableMetadata
+  )
+  {
+    final List<RelDataTypeField> sourceFields = sourceType.getFieldList();
+    for (final RelDataTypeField sourceField : sourceFields) {
+      // Check that there are no unnamed columns in the insert.
+      if (UNNAMED_COLUMN_PATTERN.matcher(sourceField.getName()).matches()) {
+        throw InvalidSqlInput.exception(
```

Comment:
could you replace these exceptions with ones communicating the `SqlNode` if its interesting/valuable....

if its a general error that doesn't matter...but this error is specific to a selected column:

```
        throw buildCalciteContextException(
            "Insertion requires columns to be named....",
            getSqlNodeFor(insert, sourceFields.indexOf(sourceField))
```
rough `sqlNodeFor` method:
```
  SqlNode getSqlNodeFor(SqlInsert insert, int idx) {
    SqlNode src = insert.getSource();
    if(src instanceof SqlSelect) {
      SqlSelect sqlSelect = (SqlSelect) src;
      SqlNodeList selectList = sqlSelect.getSelectList();
      if(idx < selectList.size()) {
      return selectList.get(idx);
      }
    }
    return src;
  }
```